home/6_1/Main.java

```diff
@@ -0,0 +1,228 @@
+
+class GreenMessage {
+    String message;
+    int counter;
+
+    GreenMessage(String message, int counter) {
+        this.message = message;
+        this.counter = counter;
+    }
+}
+
+class BlueMessage {
+    double val1;
+    double val2;
+
+    BlueMessage(double val1, double val2) {
+        this.val1 = val1;
+        this.val2 = val2;
+    }
+}
+
+class OrangeMessage {
+    String message1;
+    String message2;
+    int counter;
+    double val;
+
+    OrangeMessage(String message1, String message2, int counter, double val) {
+        this.message1 = message1;
+        this.message2 = message2;
+        this.counter = counter;
+        this.val = val;
+    }
+}
+
+class Subscription<T> {
+    private final UUID id;
+    private final Consumer<T> subscriber;
+
+    Subscription(UUID id, Consumer<T> subscriber) {
+        this.id = id;
+        this.subscriber = subscriber;
+    }
+
+    UUID getId() {
+        return id;
+    }
+
+    Consumer<T> getSubscriber() {
+        return subscriber;
+    }
+}
+
+class MessageDispatcher {
+    private final Lock lock = new ReentrantLock();
+    private final List<Subscription<GreenMessage>> greenSubscribers = new ArrayList<>();
+    private final List<Subscription<BlueMessage>> blueSubscribers = new ArrayList<>();
+    private final List<Subscription<OrangeMessage>> orangeSubscribers = new ArrayList<>();
+
+    UUID subscribeGreen(Consumer<GreenMessage> subscriber) {
+        lock.lock();
+        try {
+            UUID id = UUID.randomUUID();
+            greenSubscribers.add(new Subscription<>(id, subscriber));
+            return id;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    UUID subscribeBlue(Consumer<BlueMessage> subscriber) {
+        lock.lock();
+        try {
+            UUID id = UUID.randomUUID();
+            blueSubscribers.add(new Subscription<>(id, subscriber));
+            return id;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    UUID subscribeOrange(Consumer<OrangeMessage> subscriber) {
+        lock.lock();
+        try {
+            UUID id = UUID.randomUUID();
+            orangeSubscribers.add(new Subscription<>(id, subscriber));
+            return id;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    void unsubscribeGreen(UUID id) {
+        lock.lock();
+        try {
+            greenSubscribers.removeIf(sub -> sub.getId().equals(id));
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    void unsubscribeBlue(UUID id) {
+        lock.lock();
+        try {
+            blueSubscribers.removeIf(sub -> sub.getId().equals(id));
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    void unsubscribeOrange(UUID id) {
+        lock.lock();
+        try {
+            orangeSubscribers.removeIf(sub -> sub.getId().equals(id));
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    void publishGreen(GreenMessage message) {
```

Comment:
We have a potential issue with publishing and handling of messages. It is better to publish messages instantly and return from the function while the other thread will dispatch them and will call corresponding handlers. At the moment the thread that publishes the message can hang for some long time if we have too many handlers or one of them takes a long time. And it may be a problem - let's suppose your publisher is GUI which reports that some button was pressed - if handler of that message will be executed for 1-2 seconds then the GUI thread should wait too which may produce a bad user experience.